### DISCUSSION TITLE
Model output attributes table error

### DISCUSSION THREAD

**Discussion by evanr1232** [NASA-LIS/LISF#1378] [Q&A] [answered]
Hello,

I am currently trying to generate a LIS spinup file, but am having trouble with the model output attributes file when running LIS:

[ERR] Qrec                                               field is not defined
[ERR] for diagnostic output...
[ERR] Please exclude it from the model output attributes table

From reading previous discussion posts, I probably have the grip_id wrong or similar:

Qrec:         1  kg/m2s  IN   1 0 0 1 163 10000   # Recharge (kg/m2s)

The rest of the file, if that is helpful (converted to .txt):
[output_table.txt](https://github.com/NASA-LIS/LISF/files/12021636/output_table.txt)

Any help would be greatly appreciated! Thank you!

**Comment by dmocko** [comment, marked as answer]
Hi @evanr1232 

What LSM are you using?  Very few of the current LSMs in LIS have "Qrec" available as an output field.  That is why you see that error.

If you are using Noah-MP (version 3.6 or version 4.0.1), there are two separate variables for recharge available.  One giving recharge from groundwater to the water table, and one giving recharge to the groundwater from the water table.  Add these lines to your MODEL OUTPUT TBL file:

```
RechToGW:       1 m - 1 0 0 1 164 10000 # recharge to the water table when groundwater is deep
RechFromGW:     1 m - 1 0 0 1 165 10000 # recharge from the water table when groundwater is shallow
```

Unless you are writing your output in GRIB format, the last two columns of numbers (e.g., 164 10000) aren't used at all.

Note that the units of these two variables must be meters, which is why you have to list them as "m".

Change the first column/number for the "Qrec:" line to "0" as well.

**Reply by evanr1232** [reply]
Hi @dmocko,

Thank you for your reply! This fixed the issue.

I am currently using Noah-MP 4.0.1. This [users guide](https://lis.gsfc.nasa.gov/sites/default/files/LIS/public_7_3_releases/LIS_usersguide.21-dec-2020.html#_model_output_specifications) mentions to check the source code to see all of the possible output fields for a model - I was wondering if you could point me to the source code that specifies the outputs for Noah-MP 4.0.1, as I can't seem to find it?

Thank you!

**Reply by dmocko** [reply]
Yes, for Noah-MP-4.0.1, you would look at this file:

https://github.com/NASA-LIS/LISF/blob/master/lis/surfacemodels/land/noahmp.4.0.1/NoahMP401_main.F90

Search the code for `LIS_MOC_`.  These lines of code list possible output variables from this LSM.  There may be repeated lines, to give different potential units (that you would select in your MODEL OUTPUT TBL).  For example, see `LIS_MOC_RHMIN` - there are two units available: either unitless `-` or percentage `%`.

For other LSMs, look for the `_main.F90` code in that LSM's directory and again search for `LIS_MOC_`.

**Reply by sudarsan3010** [reply]
Hi @dmocko ,

I have recently gone through [Noah-MP's technical notes](https://opensky.ucar.edu/islandora/object/technotes:599).
The meaning/definition of the recharge variables are given as follows,

groundwater recharge to or from the water table when deep
groundwater recharge to or from the water table when shallow

What you have referred above is something different from this or the same? 

"recharge to the water table when groundwater is deep
recharge from the water table when groundwater is shallow"

If so, the code in LISF's Noah-MP-4.0.1 is modified accordingly or same as the original Noah-MP codes?

**Reply by dmocko** [reply]
HI @sudarsan3010 

Between Noah-MP-4.0.1 and Noah-MP-5.0, the NCAR maintainers of the Noah-MP LSM did a complete re-factoring/re-write of the code, including updated variable names and inline documentation.

For Noah-MP-4.0.1, if you want to output these variables, you need to change the first column after the variable name to "1" for:

```
RechToGW:           1  m       -    1 0 0 1 148 10      # recharge to water table when groundwater is deep
RechFromGW:         1  m       -    1 0 0 1 148 10      # recharge from water table when groundwater shallow
```

`RechToGW` is the same as `DEEPRECH`, and `RechFromGW` is the same as `RECH`.
